ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineContext.java

```diff
@@ -227,7 +227,7 @@ void registerQuery(final QueryMetadata query) {
     }
   }
 
-  private void unregisterQuery(final QueryMetadata query) {
+  void unregisterQuery(final QueryMetadata query) {
```

Comment:
I think we should piggy-back on the work in #5766 for query upgrades; we could consider a restart as an upgrade which just upgrades to itself. That way we could reuse the "replace query" codepath that I worked on for that - specifically if you call `registerQuery` with a query with the same ID it will close the old one and register the new one. Also `unregister` remains private, which I think is an important abstraction detail

Reply:
Interesting - what would that look like? We'd need some API in `QueryMetadata` for cloning the existing query. And then we'd pass the cloned query to `registerQuery`?

Reply:
yup, as long as it has the same `queryId` it would handle replacing (and closing) the old one. the big benefit is that we get to harden both paths with one go (issues such as https://github.com/confluentinc/ksql/pull/5807#discussion_r455455174 which I didn't think about when implementing query upgrades)

Reply:
I'm not sure if the `EngineContext.registerQuery` will work @agavra. 

Looking at it, it does two things that should not happen:
- calls oldQuery.close() which cleans-up the state directory (we need to keep this state directory on restarts)
- updates the metastore source and sink names (it will fail if the source/sink are already registered)

Should the metastore.updateForPersistentQuery() fail in the above case? Can I refactor it to continue if the source and sink are already there? 

I could avoid calling the `unregisterQuery` to prevent the clean by refactoring the `registerQuery` to optional clean-up the state directory and do not fail on updating the metastore. One thing, why does the `registerQuery` closes an old query instead of failing if the queryID is already registered? Shouldn't that be the case like the metastore update?

Reply:
> calls oldQuery.close() which cleans-up the state directory (we need to keep this state directory on restarts)

Reply:
@agavra I refactored the code and moved the restart to the `PersistentQueryMetadata`. Instead of unregistering/registering the query or piggyback on the query upgrades, I just made sure to restart the internal Kafka streams. 

I had to refactor some code to make it work, like:
- encapsulating the creation of the Kafka streams in the QueryMetadata and PersistentQueryMetadata (for restarts).
- re-create the materialization provider in the PersistentQueryMetadata on restarts.
- pass the KafkaStreamsBuilder from the QueryExecutor to the QueryMetadata for building KafkaStreams